larpandoradlcontent/LArSignalId/DlSNSignalAlgorithm.cc

```diff
@@ -0,0 +1,1060 @@
+        {
+            std::cout << "SignalAssessmentAlgorithm: Unable to write to ROOT tree" << std::endl;
+        }
+    }
+}
+
+//-----------------------------------------------------------------------------------------------------------------------------------------
+
+StatusCode DlSNSignalAlgorithm::Run()
+{
+    if (m_visualise)
+    {
+	PANDORA_MONITORING_API(SetEveDisplayParameters(this->GetPandora(), true, DETECTOR_VIEW_XZ, -1.f, 1.f, 1.f));
+    }
+    
+    if (!m_applyCheatedSeparation)
+    {    
+        if (m_trainingMode)	    
+	    return this->PrepareTrainingSample();
+        else
+            return this->Infer();
+    }
+    else
+        return this->CheatedSeparation();
+
+    return STATUS_CODE_SUCCESS;
+}
+
+StatusCode DlSNSignalAlgorithm::PrepareTrainingSample()
+{
+    LArMCParticleHelper::MCContributionMap mcToHitsMap;
+    PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, this->GetMCToHitsMap(mcToHitsMap));
+    if (mcToHitsMap.empty())
+        throw StatusCodeException(STATUS_CODE_NOT_ALLOWED);
+
+    // Get boundaries for hits and make x dimension common
+    std::map<int, float> wireMin, wireMax;
+    std::map<int, bool> viewCalculated;
+    float driftMin{std::numeric_limits<float>::max()}, driftMax{-std::numeric_limits<float>::max()};
+    for (const std::string &listname : m_inputCaloHitListNames)
+    {
+        const CaloHitList *pCaloHitList{nullptr};
+	try
+        {
+            PandoraContentApi::GetList(*this, listname, pCaloHitList);
+	}
+	catch (const StatusCodeException &e)
+	{
+            continue;
+	}
+        if (!pCaloHitList || pCaloHitList->empty())
+            continue;
+
+        HitType view{pCaloHitList->front()->GetHitType()};
+        float viewDriftMin{driftMin}, viewDriftMax{driftMax};
+        try
+        {
+            this->GetHitRegion(*pCaloHitList, viewDriftMin, viewDriftMax, wireMin[view], wireMax[view]);
+	    viewCalculated[view] = true;
+        }
+        catch (const StatusCodeException &e)
+        {
+            if (e.GetStatusCode() == STATUS_CODE_NOT_FOUND)
+	        continue;
+	    else
+                throw;
+        }
+
+	driftMin = std::min(viewDriftMin, driftMin);
+        driftMax = std::max(viewDriftMax, driftMax);
+    }
+
+    for (const std::string &listName : m_caloHitListNames)
+    {
+        const CaloHitList *pCaloHitList{nullptr};
+        PANDORA_RETURN_RESULT_IF_AND_IF(STATUS_CODE_SUCCESS, STATUS_CODE_NOT_INITIALIZED, !=, PandoraContentApi::GetList(*this, listName, pCaloHitList));
+
+        if (!pCaloHitList)
+        {
+            std::cout << "ERR: Could not find full CaloHitList - DlSNSignalAlgorithm unable to proceed" << std::endl;
+            continue;
+        }
+
+        HitType view{pCaloHitList->front()->GetHitType()};
+        float viewDriftMin{driftMin}, viewDriftMax{driftMax};
+        if (!viewCalculated[view])
+        {
+            const LArTPC *const pTPC(this->GetPandora().GetGeometry()->GetLArTPCMap().begin()->second);
```

Comment:
Just to clarify, this only ever returns the first LArTPC in the map.  Does that matter if the hits are contained in a different volume (where the views might be swapped)?